### ISSUE TITLE
PingingExceptions and SocketExceptions on dead node not surfacing

### ISSUE THREAD

**Issue by mjohnst** [elastic/elasticsearch-net#1503]
In our ES cluster, we had a node that was offline intentionally, but we had forgotten to remove the associated Uri from our list. It was still being considered when the client was selecting the next node for requests.

We weren't receiving any indication that this node was dead. It only became apparent to us when we added some additional logging to the RetryRequestAsync method to attempt to catch a separate issue.

Every minute or so, we were getting this exception:
```c#
Elasticsearch.Net.Exceptions.PingException: Pinging http://192.168.0.4:9200/ caused an exception
   at Elasticsearch.Net.Connection.Transport.<>c__DisplayClass4.<Elasticsearch.Net.Connection.ITransportDelegator.PingAsync>b__2(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()" TraceSource="w3wp.exe"
```

Followed by another exception every 10 or so minutes:
```c#
System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond 192.168.0.4:9200
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.EndGetRequestStream(IAsyncResult asyncResult, TransportContext& context)
   at System.Net.HttpWebRequest.EndGetRequestStream(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)" TraceSource="w3wp.exe"
```

After we fixed the issue (reviving the dead node) the exceptions stopped.

These exceptions should be surfaced more loudly so that we can be aware when a node is unresponsive.

**Comment by Mpdreamz** [comment]
Thats  very odd, will have to deep dive why this is happening. The ping should have put that node in the dog house and not be electable until a ping succeeds on it. The way that works is that when a ping fails its put in the dog house for a duration of `N`, after which it becomes available again but with a flag indicating a ping is required for first usage of this node again, if that still fails its put in the doghouse for `N * F`, all the way up to `MIN(M, N* F)`. 
`M` and `N` and are configurable here: http://nest.azurewebsites.net/elasticsearch-net/cluster-failover.html

I wonder if it has anything to do with how you reseed the cluster on an interval as mentioned here:
https://github.com/elastic/elasticsearch-net/issues/1499#issuecomment-125374491 which would clear this administration. Not trying to passing blame as I still think the mechanism should have worked. 

We have a pretty extensive test suite for this stuff see e.g [here](https://github.com/elastic/elasticsearch-net/blob/develop/src/Tests/Elasticsearch.Net.Tests.Unit/Failover/Revival/RevivalTests.cs#L98) 

I would love to try and replicate the behaviour you are seeing in one of them. Can you share some information about the topology (N nodes), where were you seeing this exception? was it thrown or passed as information on the response ?

**Comment by Mpdreamz** [comment]
Just as FYI we will change this in 2.0 a little bit, metrics will be gathered by default. 

Thinking about your request to surface them more loudly we can add a `FellOver` bool to all the responses. This allows you to write a generalised logger/notifier a whole lot easier.

**Comment by mjohnst** [comment]
Our topology is Azure hosted nodes with data, master, and http endpoints. We receive a list of these endpoints using our sniffer and only use the http endpoints. We load-balance between the valid endpoints using a round robin method.

We discovered the pings when I added some logging to the `RetryRequestAsync<T>(TransportRequestState<T> requestState)` method when we were trying to fix #1499 in our environment:

```c#
if (requestState.SeenExceptions != null)
    System.Diagnostics.Trace.TraceWarning(
        string.Format("Retrying a async request.{0}RetryCount: {1}{0}Waiting {2}ms before retry.{0}Current seen exceptions:{0}{3}",
            Environment.NewLine, requestState.Retried, requestState.Retried * 100, string.Join("\n", from e in requestState.SeenExceptions select e.ToString())));
else
    System.Diagnostics.Trace.TraceWarning(
        string.Format("Retrying a async request.{0}RetryCount: {1}{0}Waiting {2}ms before retry.{0}Current seen exceptions is null.",
            Environment.NewLine, requestState.Retried, requestState.Retried * 100));

int retrySleepFactor = 1000;
System.Threading.Thread.Sleep(requestState.Retried * retrySleepFactor);
```

Both the socket and pinging exceptions were being traced. The exceptions (both pinging and socket) had `requestState.Retried` equal to 0 when they hit our trace.

Let me know if you need more details to repro.

**Comment by Mpdreamz** [comment]
Hi @mjohnst also closing this issue because its old. If you are still seeing this we would love to help you diagnose it again. 

Again our 2.0 branch has received a lot of love and has a much much saner exception approach.
